Mathlib/RingTheory/MvPolynomial/NewtonIdentities.lean

```diff
@@ -0,0 +1,327 @@
+
+As in other polynomial files, we typically use the notation:
+
++ `σ τ : Type _` (indexing the variables)
+
++ `R S : Type _` `[CommSemiring R]` `[CommSemiring S]` (the coefficients)
+
++ `r : R` elements of the coefficient ring
+
++ `i : σ`, with corresponding monomial `X i`, often denoted `X_i` by mathematicians
+
++ `φ ψ : MvPolynomial σ R`
+
+## References
+
+See [zeilberger1984] for the combinatorial proof of Newton's identities.
+-/
+
+open Equiv (Perm)
+
+open BigOperators MvPolynomial
+
+noncomputable section
+
+section PowerSum
+
+namespace Multiset
+
+variable {R : Type _} [CommSemiring R]
+
+/-- The degree-`n` power sum evaluated at the elements of `s` -/
+def psum (s : Multiset R) (n : ℕ) : R := (s.map (fun (x : R) ↦ x ^ n)).sum
+
+end Multiset
+
+open Finset
+
+variable (σ R) [CommSemiring R] [Fintype σ] [Fintype τ]
+
+/-- The degree-`n` power sum -/
+def psum (n : ℕ) : MvPolynomial σ R := ∑ i, (X i) ^ n
+
+theorem rename_psum (n : ℕ) (e : σ ≃ τ) : rename e (psum σ R n) = psum τ R n := by
+  calc
+    rename e (psum σ R n) = ∑ i, (X (e i)) ^ n := by
+      simp_rw [psum, map_sum, map_pow, rename_X]
+    _ = ∑ i in univ.map e.toEmbedding, (X i) ^ n := by simp [-map_univ_equiv]
+    _ = ∑ i, (X i) ^ n := by rw [map_univ_equiv]
+
+theorem psum_isSymmetric (n : ℕ) : IsSymmetric (psum σ R n) := by
+  intro
+  rw [rename_psum]
+
+end PowerSum
+
+namespace Newton
+
+open Finset Nat
+
+variable (σ : Type _) [Fintype σ] [DecidableEq σ] [Fintype τ] (R : Type _) [CommRing R]
+  [NoZeroDivisors (MvPolynomial σ R)] [CharZero (MvPolynomial σ R)]
+/-
+  TODO: show that MvPolynomial σ R is an integral domain if R is an integral domain
+  TODO: show that MvPolynomial σ R has characteristic zero if R has characteristic zero
+-/
+
+-- The following proof is from Zeilberger, "A combinatorial proof of Newton's identities" (1984)
+def PairsPred (k : ℕ) (t : Finset σ × σ) : Prop := card t.fst ≤ k ∧ (card t.fst = k → t.snd ∈ t.fst)
+
+open Classical in
+def pairs (σ : Type _) [Fintype σ] (k : ℕ) : Finset (Finset σ × σ) :=
+  Finset.univ.filter (PairsPred σ k)
+
+open Classical in
+def card_eq_if_not_lt (t : Finset σ × σ) (ht : t ∈ pairs σ k) (hnlt : ¬card t.fst < k) :
+    card t.fst = k := by
+  simp_rw [pairs, mem_filter, PairsPred] at ht
+  exact Or.resolve_right (le_iff_eq_or_lt.mp ht.2.1) hnlt
+
+def weight (k : ℕ) (t : Finset σ × σ) : MvPolynomial σ R :=
+  (-1) ^ (card t.fst) * ((∏ a in t.fst, X a) * (X t.snd) ^ (k - card t.fst) : MvPolynomial σ R)
+
+def T_map (t : Finset σ × σ) : Finset σ × σ :=
+  if t.snd ∈ t.fst then (t.fst.erase t.snd, t.snd) else (insert t.snd t.fst, t.snd)
+
+/-- Needed for `Finset.sum_involution` -/
+def T_map_restr (t : Finset σ × σ) (_ : t ∈ pairs σ k) := T_map σ t
+
+open Classical in
+theorem T_map_pair (t : Finset σ × σ) (h : t ∈ pairs σ k) : T_map_restr σ t h ∈ pairs σ k := by
+  rw [pairs, mem_filter, PairsPred] at *
+  simp_rw [T_map_restr, T_map]
+  split_ifs with h1
+  · simp_all
```

Comment:
You'll need to remove all the `simp_all` and `simp` calls that don't close the goal (replace by `simp only [...]`)

Reply:
Will go through and find these with tracing of simps enabled.

Reply:
Okay, I think the only remaining `simp`s (excepting `simp only`s), `simp_all`s, and `simp_rw`s directly close goals.